### PR TITLE
Plot PDOS from CASTEP

### PR DESCRIPTION
**Pull request by zhubonan**
@ajjackson I have added the ability to plot CASTEP PDOS. The actual binary file is read using [castepxbin](https://github.com/zhubonan/castepxbin), which also does some processing such as assigning the angular moment channel to orbitals and orbitals sites.

A few things:

- I am not sure how the f orbitals in CASTEP relates to that in VASP. The latter does not have any meaningful labels, so I just naively assigned them in order they appear.
- I have only tested with the `Si2` example found in the CASTEP's tutorial. The generated PDOS plot is consistent with that of from OptaDos. 
- In the future, the dependency `castepxbin` will be released to pypi, I added its repo address and pinned to `v0.0.1` tag in the `setup.py` for now.

Si2 Example:
Left - OptaDOS
Right - Sumo
image

### Testing instruction

First clone my repository (ideally in a new venv so there is no risk of messing up the current installation):

```
git clone https://github.com/zhubonan/sumo.git
```

Then checkout the `pr-castep-pdos` branch containing the update.

```
cd sumo
git checkout pr-castep-pdos
```

Once this is done, install the local repository using `pip`:

```
cd ../
pip install -e ./sumo
```

Afterwards, do your usual `sumo-dosplot -c castep -g 0.1`, it will print something like:

```
Found PDOS binary file Si2.pdos_bin, including PDOS in the plot.  
Reading CASTEP .bands file header...
nbands: 23, efermi / Ry: 0.165228, spin channels: 1
Reading band eigenvalues...
Electronic structure appears to have a bandgap
Setting energy zero to VBM
```

### CODE REVIEW

**Comment by sajidurehman** [comment]
Hi zhubonan,
                 I am interested to use the sumo to plot the CASTEP PDOS. I again downloaded the sumo from here and again reinstall. But i am unable to get the PDOS plot. I used the following command.

sumo-dosplot --code castep -f Si2.bands -g 0.1

Could you please let me know how to properly use the sumo for CASTEP PDOS. Looking forward to here you. Thank you in advance.

**Comment by zhubonan** [comment]
Hi @sajidurehman, glad to know there are people interested in this functionality. 

Are you sure you installed from my fork? This PR is not merged into the official repository yet, so you cannot use the standard installation instruction.

You can install my fork with:

```
pip install git+https://github.com/zhubonan/sumo.git@pr-castep-pdos
```
Note that is only a temporary solution, once this PR is merged the link above will no longer work. But hopefully, it gives you a sneak peek of the capability. 

On my machine, this works with the Si2 example, using:

```
sumo-dosplot -c castep  -g 0.1
```

but with the `-f Si2.bands` option, it does give an error. So I need to fix it! I am not so familiar with the `sumo` codebase yet.


If it still not work for you, please do post the error message you get and I will investigate further.

**Comment by sajidurehman** [comment]
Hi zhubonan,
             Thank you for reply and PDOS feature. After your suggestion I successfully installed your fork. Now both commands working smoothly without any error. 
sumo-dosplot --code castep -f Si2.bands -g 0.1
sumo-dosplot -c castep  -g 0.1

But still getting only total DOS.  Do I need any additionaly linkage and steps for proper usage.

**Comment by zhubonan** [comment]
@sajidurehman Do you have both the `Si2.pdos_bin` and `Si2.cell` present in the plotting folder? It will fallback to total DOS only if the `Si2.pdos_bin` is not there.

**Comment by sajidurehman** [comment]
Yes, both files are there in same folder. 
image

**Comment by zhubonan** [comment]
@sajidurehman I think this is a installation problem. If you already have `sumo` in your python installation pip will refuse to overwrite... Please use the instructions below to properly checkout the pr code. 

First clone my repository:

```
git clone https://github.com/zhubonan/sumo.git
```

Then checkout the `pr-castep-pdos` branch containing the update.

```
cd sumo
git checkout pr-castep-pdos
```

Once this is done, install the local repository using `pip`:

```
cd ../
pip install -e ./sumo
```

Note that I use the `-e` flag to install it in the development mode, so that any change in the local repo will have effect without the need for installation. 

Once this is done, do your usual `sumo-dosplot -c castep -g 0.1`, it will print something like:

```
Found PDOS binary file Si2.pdos_bin, including PDOS in the plot.  
Reading CASTEP .bands file header...
nbands: 23, efermi / Ry: 0.165228, spin channels: 1
Reading band eigenvalues...
Electronic structure appears to have a bandgap
Setting energy zero to VBM
```

If you don't see the first line saying something about PDOS, you are not running with the code in this PR.

**Comment by zhubonan** [comment]
@sajidurehman Do let me know if it works for you. Feedbacks (and bugs) are very welcome!

**Comment by sajidurehman** [comment]
Thank you so much for detailed steps. Now scuccessfully obained PDOS of Si. Then I tried on GaN. Ga has three orbitals s,p,d and N has two orbital s,p. But the obtained dos.jpg missing some orbital.
image
is this some bug or plotting issue?

**Comment by zhubonan** [comment]
@sajidurehman Thanks! That sounds promising. 

I think maybe this is a plotting problem. I can see there is a cyan line which has no legend assigned. 
Perhaps you can try requesting the channel explicitly, or just increase the plotting range? I found somo tries to avoid printing the legend if there is not much density of the channel in the field of view. 

If that does not work, or you find the plot is just wrong. Please do let me have the input files and I will investigate further.

**Comment by sajidurehman** [comment]
Great! you are absolutely right. I increased the range -20 to 20 then got all otbitals.

I have some confusion regarding PDOS of sumo and optados results. Both have different results from same calculation why ?

**Comment by zhubonan** [comment]
@sajidurehman Can you please give me an example?

The one from OptaDos may have been processed with some advanced broadening techniques such as the *adaptive* broadening scheme, or with a different smearing width. The one you get from `sumo` is just what CASTEP throws out.  So it is natural that the two are not the same, especially if your kpoint grid is not fully converged. But at least the line shapes should be similar. 

The other thing to note is that `sumo` can output *lm* resolved orbitals if you request it, for example, using `-o Si.p`.

**Comment by sajidurehman** [comment]
I tried the GaN as an example. Please see the attached figure. If you notice around Fermi level both have different shape.
image

**Comment by zhubonan** [comment]
@sajidurehman Yes this is normal. You are using a different broadening scheming in OptaDos (I guess it is the default - adaptive broadening, evident by the sharp peaks and flat shoulders).
 
You can try to run another `OptaDos` calculation with the following added to your `odi` file:

```
broadening: fixed
fixed_smearing: 0.1 
```

Which make `OptaDos` uses the same fixed-width gaussian broadening as that in `sumo`.

**Comment by sajidurehman** [comment]
Thank you for guidance. 
Do you have any plan to introduce projected Band structure feature in sumo for CASTEP?

**Comment by zhubonan** [comment]
I would like to see this PR merged first (@ajjackson ), before moving on to implementing the projected band structure. It should not be too much extra work.

**Comment by sajidurehman** [comment]
Hopefully it will be implemented soon...many other people will also be benefited from this great feature.

**Comment by ajjackson** [comment]
This is looking promising! I do hope to review this soon but am trying to close out a couple of other things elsewhere.

By default Sumo does not label PDOS lines with visible intensity below 3% of the total DOS. You can set the threshold with the `--legend-cutoff` argument.

**Comment by sajidurehman** [comment]
Hi @zhubonan 
                     I am unable to generate the band structure by following command.
                     sumo-bandplot --code castep -f Si2.bands
Error:
Traceback (most recent call last):
  File "/home/max/.local/bin/sumo-bandplot", line 11, in <module>
    load_entry_point('sumo', 'console_scripts', 'sumo-bandplot')()
  File "/home/max/.local/lib/python3.6/site-packages/pkg_resources/__init__.py", line 490, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/home/max/.local/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2859, in load_entry_point
    return ep.load()
  File "/home/max/.local/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2450, in load
    return self.resolve()
  File "/home/max/.local/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2456, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/home/max/sumo/sumo/cli/bandplot.py", line 33, in <module>
    from sumo.io.castep import read_tdos as read_castep_tdos
ImportError: cannot import name 'read_tdos'

Could you please help me how to fix this error.

**Comment by zhubonan** [comment]
@ajjackson Thanks!

**Comment by zhubonan** [comment]
@sajidurehman You should update your local `sumo` copy using `git pull`, I already fixed it in e1393ce.

**Comment by sajidurehman** [comment]
Thank you so much. Now successfully got band structure.  
Could you please help me more regarding following command.
sumo-kgen -c castep -p Si2.cell

Error:
Structure information:
	Space group number: 227
	International symbol: Fd-3m
	Lattice type: cubic

k-point path:
	\Gamma -> L -> W -> X -> \Gamma

k-points:
	\Gamma: 0.0 0.0 0.0
	L: 0.5 0.5 0.5
	W: 0.5 0.25 0.75
	X: 0.5 0.0 0.5

k-point label indices:
	\Gamma: 1
	L: 62
	W: 112
	X: 147
	\Gamma: 217
Traceback (most recent call last):
  File "/home/max/.local/bin/sumo-kgen", line 11, in <module>
    load_entry_point('sumo', 'console_scripts', 'sumo-kgen')()
  File "/home/max/sumo/sumo/cli/kgen.py", line 307, in main
    phonon=args.phonon)
  File "/home/max/sumo/sumo/cli/kgen.py", line 164, in kgen
    kpath.prim.to(filename=prim_filename)
  File "/home/max/.local/lib/python3.6/site-packages/pymatgen/core/structure.py", line 2174, in to
    raise ValueError("Invalid format: `%s`" % str(fmt))
ValueError: Invalid format: ``

**Comment by ajjackson** [comment]
I've had an initial play and it seems to work for me. I notice that the `--total-only` option is not currently respected, we should ignore the PDOS data if that is present.

I also notice that `--no-shift` seems to be broken for CASTEP in general so I'd better take a look at that 😞 

Will run something with d orbitals and check that the orbital selection stuff is working.

File: setup.py
Comment (ajjackson) [inline setup.py:45]: Can we move this into "extras" rather than have it as a mandatory requirement? Maybe we can have a 'castep' option for all castep-related extras.

File: sumo/cli/bandplot.py
Comment (ajjackson) [inline sumo/cli/bandplot.py:311]: What happens if the user has ".cell" in their seedname? (e.g. `prim.cell.2.cell, prim.cell.2.param`). Would be good to make this explicitly substitute the file extension to be safe.

File: sumo/cli/bandplot.py
Comment (ajjackson) [inline sumo/cli/bandplot.py:314]: s/exis/exist

File: sumo/cli/bandplot.py
Comment (ajjackson) [inline sumo/cli/bandplot.py:318]: s/exis/exist

File: sumo/cli/bandplot.py
Comment (ajjackson) [inline sumo/cli/bandplot.py:321]: This is not PEP8 indentation

File: sumo/cli/bandplot.py
Comment (ajjackson) [inline sumo/cli/bandplot.py:324]: I think this option needs tweaking to respect user args. Not your fault, was already there!

File: sumo/cli/dosplot.py
Comment (ajjackson) [inline sumo/cli/dosplot.py:177]: Maybe we should add some hook function so whenever a TDOS is being plotted we warn the user about irrelevant arguments they asked for. That's not CASTEP-specific and should maybe be a follow-up rather than part of this PR? In the meantime it does make sense to cut this as it's no longer a CASTEP issue.

File: sumo/cli/dosplot.py
Comment (ajjackson) [inline sumo/cli/dosplot.py:190]: Again, not sure if replace() is quite robust enough

File: sumo/cli/dosplot.py
Comment (ajjackson) [inline sumo/cli/dosplot.py:202]: f-strings are usable as of Python 3.6. Officially v1.3.0 is the last Sumo version to support python 3.5, but setup.py still supports it. So if this is the first PR to include f-strings, I think it is also time to update setup.py and Travis to actively drop 3.5! @utf is that ok?

File: sumo/io/castep.py
Comment (ajjackson) [inline sumo/io/castep.py:20]: Realistically this import is going to fail sometimes. I guess our options are:

1. make this a fully optional dependency, and fall back to TDOS with a warning message if it is not installed
2. make this a required dependency _for CASTEP users_ and provide a more helpful error message if this import fails
3. make this part of the standard sumo installation (as you have done so far in the PR). In which case the default error is ok

My current preference would be (1) but I'd be ok with (2) if the castepxbin package is on PyPI. (Not keen on github imports in setup.py files.) I remain open to being persuaded that (3) is the way forward especially if @utf agrees.

File: sumo/cli/dosplot.py
Comment (ajjackson) [reply sumo/cli/dosplot.py:202]: (This is not the only f-string in the change set, but it's the first one I "noticed" as I have also been using them a lot on other projects 😁 )

**Comment by zhubonan** [comment]
@ajjackson Thanks for the comments. I will work through them soon!

File: sumo/cli/dosplot.py
Comment (zhubonan) [reply sumo/cli/dosplot.py:177]: Probably better for another PR to address it once for all? I think it is better to be addressed at a code agnostic level.

File: setup.py
Comment (zhubonan) [reply setup.py]: I have updated it - now `castepxbin` is release on pypi. I have pinned it to the `0.1.0` version for now.

File: sumo/cli/bandplot.py
Comment (zhubonan) [reply sumo/cli/bandplot.py]: solved with more robust method for file extension substitution

File: sumo/io/castep.py
Comment (zhubonan) [reply sumo/io/castep.py:20]: (3) can be taken as now the external code is released on pypi

File: sumo/cli/dosplot.py
Comment (zhubonan) [reply sumo/cli/dosplot.py]: I have removed f-string usage for now, since v1.4.0 still support python 3.5

File: sumo/cli/bandplot.py
Comment (zhubonan) [reply sumo/cli/bandplot.py]: done

File: sumo/cli/bandplot.py
Comment (zhubonan) [reply sumo/cli/bandplot.py]: I guess you mean the `efermi_to_vbm`? There seem to be no options about it in the outer function...

Comment (ajjackson) [reply sumo/cli/bandplot.py]: Yup, I think an overhaul of Fermi level handling in Sumo needs to be follow-up work on this. Let's get this PR done first.

File: sumo/cli/dosplot.py
Comment (ajjackson) [reply sumo/cli/dosplot.py:177]: Created issue #26 to record this

File: sumo/cli/dosplot.py
Comment (ajjackson) [reply sumo/cli/dosplot.py]: Now that we are on release 2.0 we can make a release that officially breaks Python 3.6 and unlock f-strings. I'd rather close this PR before then, so thanks for changing.

File: sumo/io/castep.py
Comment (ajjackson) [reply sumo/io/castep.py:20]: As castepxbin is a small package without onerous dependencies I am happy to accept it as a mandatory dependency of Sumo at this stage. However, I note that this approach will not be suitable for all code interfaces; if we start using ASE that should probably be optional.

**Comment by ajjackson** [comment]
Ok, I have had a play with a nasty spin-polarised NiO case and the dosplot options seem to work as expected :smile: . Fairly happy to accept this now, but would like to have some kind of testing in place. I guess the best way to do this would be to mock the the external function `compute_pdos`, checking that it is passed the correct information and that the return values are correctly handled in Sumo. Any thoughts?

Ideally we would also have some kind of "integration test" with real data that can be run separately / less frequently. That sounds like a separate project.

**Comment by ajjackson** [comment]
There seems to still be an issue with `--total--only` not working.

**Comment by zhubonan** [comment]
> 
> 
> Ok, I have had a play with a nasty spin-polarised NiO case and the dosplot options seem to work as expected 😄 . Fairly happy to accept this now, but would like to have some kind of testing in place. I guess the best way to do this would be to mock the the external function `compute_pdos`, checking that it is passed the correct information and that the return values are correctly handled in Sumo. Any thoughts?
> 
> Ideally we would also have some kind of "integration test" with real data that can be run separately / less frequently. That sounds like a separate project.

I don't think there is a need to mock `compute_dos` ? `castepxbin` should be installed automatelly in the test enviroment as well. Just add use some real data for the fully integration test would be nice.

**Comment by zhubonan** [comment]
@ajjackson `--total-only` should work now, did not pay much attension to that code path!
Ideally as a CLI packages it would be great to do integration test with CLI. Not sure how to do that easily though. I know `click` has a nice testing interface, but there seems to be no off-the-shelve solutions for `argparse`?

**Comment by ajjackson** [comment]
So far the approach in Sumo has been to have as much parity as reasonable practical between the argparse interface and the main function (with a similar name to CLI tool) of each sumo.cli submodule. Then that function can be called for a reasonable emulation of the command-line tool.

I don't mind running integration tests on the Sumo side as long as they aren't too slow. But I'm worried about the growing repository size with test data. Is there a way to get a fairly small pdos file? My 4-atom NiO PDOS came in at about 40MB, which isn't very git-friendly. Most of the other test data is a few MB at most.

**Comment by zhubonan** [comment]
> My 4-atom NiO PDOS came in at about 40MB, which isn't very git-friendly. Most of the other test data is a few MB at most.

Thats quite a lot... Maybe try with fewer kpoints? You don't have to use a well converged PDOS for the sake of testing 😄

**Comment by ajjackson** [comment]
Probably a good idea, although I'm not sure you can really call it a DOS with only a few k-points :wink:

**Comment by zhubonan** [comment]
just go gamma only with `spectral_kpoint_mp_grid: 1 1 1`

**Comment by sajidurehman** [comment]
Hi @zhubonan, I hope you would be find. Now sumo 2.1.1 is relased. Do you have any plan for projected band structure for CASTEP code. I am axiously waiting for this exciting feature in sumo.

Looking forward to hear you

**Comment by ajjackson** [comment]
This will happen when I stop being rubbish and remember to work on it. I'm on holiday now so maybe that will help ;-)

**Comment by sajidurehman** [comment]
Thank you for reply and considering the request.  Happy Holidays !!!

**Comment by ajjackson** [comment]
@utf I'm not seeing a Gihub Action kick in to run tests on this PR. Is there something we need to do to enable it, or are they just slow sometimes?

**Comment by ajjackson** [comment]
Ok, I have added some regression tests and managed to get the reference data down to a reasonable size; I used a 2x2x2 k-point mesh and also added gzip compression for the Pymatgen MSON files. (Hopefully these will not need to be updated many times...)

**Review by ajjackson — APPROVED** [review]
As discussed, this is a very welcome set of changes. I approve now, but we should not merge until we have seen the tests run successfully on the CI.

(I will also make a small documentation update which, again, would be good to preview somehow.)

**Comment by ajjackson** [comment]
Argh. I have tried to change the GH actions to trigger on external PRs like this one, but it still doesn't seem to work. Think I'd better wait for @utf to have a look before I push any more apparently-useless changes to master.

**Comment by zhubonan** [comment]
I think `on: [push, pull_request]` would just work fine?

**Comment by zhubonan** [comment]
The alternative is if @ajjackson create a branch under `SMTG-UCL/sumo` and pull the branch from my fork, then push. It should trigger the CI.

**Comment by sajidurehman** [comment]
Thank you @ajjackson @zhubonan for adding the pdos feature for CASTEP code. Its working smoothly. Once again thank you.

Now waiting for more exciting projected band structure feature for CASTEP code.

**Pull request merged**
